Source/Core/DiscIO/VolumeWii.cpp

```diff
@@ -521,4 +525,127 @@ bool VolumeWii::CheckBlockIntegrity(u64 block_index, const Partition& partition)
   return CheckBlockIntegrity(block_index, cluster, partition);
 }
 
+bool VolumeWii::EncryptGroup(u64 offset, u64 partition_data_offset,
+                             u64 partition_data_decrypted_size,
+                             const std::array<u8, AES_KEY_SIZE>& key, BlobReader* blob,
+                             std::array<u8, GROUP_TOTAL_SIZE>* out)
+{
+  std::vector<std::array<u8, BLOCK_DATA_SIZE>> unencrypted_data(BLOCKS_PER_GROUP);
+  std::vector<std::array<u8, BLOCK_HEADER_SIZE>> unencrypted_hashes(BLOCKS_PER_GROUP);
+
+  std::array<std::promise<bool>, BLOCKS_PER_GROUP> read_promises;
+  std::array<std::future<bool>, BLOCKS_PER_GROUP> h0_futures;
+  std::array<std::future<bool>, BLOCKS_PER_GROUP / 8> h1_futures;
+  std::future<bool> h2_future;
+
+  for (size_t i = 0; i < BLOCKS_PER_GROUP; ++i)
+  {
+    h0_futures[i] = std::async(std::launch::async,
```

Comment:
This could launch BLOCKS_PER_GROUP threads, since std::async does not necessarily use thread pools (as far as I can tell). Since you're doing the same for h1_futures and encryption_futures, this function would launch up to 3 * 0x40 = 192 threads, which is way too much.

Reply:
Ah, I was hoping it would use some kind of thread pool :( It seems to not be using a too large amount of threads on my computer, but I can't be sure that this behavior isn't Windows-specific...

(For what it's worth, there are only 8 threads for the H1 hashes, and the encryption threads aren't launched until all the other tasks are done, so we wouldn't have more than 72 threads active at once, but that's still a relatively large amount.)

Do you know of a way to do something similar but without using too many threads? I could probably manually assign the tasks to threads if there's no other good option, but I'm not sure if we have a convenient way of knowing how many cores the system has, which would be good to use as a limit for how many threads we launch.

Reply:
Unfortunately not using a thread pool manually, as I don't think `<async>` has anything like that :/

To get the number of cores, `std::thread::hardware_concurrency()` should suffice.

Reply:
Ah, I didn't know that there was something like `std::thread::hardware_concurrency()`. I was imagining that we would need to write some code in `Common` that uses platform-dependent ifdefs...

Reply:
I've made some changes to it. The hashing part still uses as many `std::async` calls as before (well, actually one less – the H2 calculation no longer uses a future), but there are much fewer of them active at once now (because the H0 hash calculations aren't started until the relevant data has been read from the disc). The encryption part uses roughly as many `std::async` calls as there are hardware threads. With this change, Dolphin's total thread count when running verification is now the same when using re-encryption as when not using re-encryption: 26 threads. (I'm not quite sure how that works, but I guess some sort of thread pool is involved on at least Windows after all.) Is this a good enough fix for your concerns?

Reply:
I have now also reduced the number of `std::async` calls for the hashing part to 64.

Reply:
Yep, that sounds a lot better. I suppose ideally we would have a thread pool implementation in Common but that's probably out of the scope of this PR.